From 877f22a4caf2fdd09465e60d2660cc913cd0fe3b Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Fri, 4 Jul 2025 15:36:36 +0800
Subject: [PATCH 108/128] mtk: mac80211: fix direct csa issues

When csa count is set to 0, then AP will trigger the channel switch
directly, which might cause some issues.
1. For mbss case, the csa finalize process should wait until all the
interface are ready to use reserved chanctx. Otherwise, the channel
switch will fail.
2. For mld, the partner links of the affiliated mld of the channel
swicth link should also perform csa finalize in order to update the
RNR IE in their beacons.

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 net/mac80211/cfg.c | 35 +++++++++++++++++++++++++++++++++--
 1 file changed, 33 insertions(+), 2 deletions(-)

diff --git a/net/mac80211/cfg.c b/net/mac80211/cfg.c
index fd57069d..db8b55a2 100644
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -4522,8 +4522,39 @@ __ieee80211_channel_switch(struct wiphy *wiphy, struct net_device *dev,
 		ieee80211_link_info_change_notify(sdata, link_data, changed);
 		drv_channel_switch_beacon(sdata, &link_data->csa.chanreq.oper);
 	} else {
-		/* if the beacon didn't change, we can finalize immediately */
-		ieee80211_csa_finalize(link_data);
+		struct ieee80211_link_data *link;
+		int n_assigned = 0, n_reserved = 0;
+		unsigned int link_id;
+
+		/* if the beacon didn't change, we can finalize after
+		 * all assigned links are ready.
+		 */
+		for_each_sdata_link(local, link) {
+			if (rcu_access_pointer(link->conf->chanctx_conf) == &chanctx->conf)
+				n_assigned++;
+			if (link->reserved_chanctx == chanctx)
+				n_reserved++;
+		}
+
+		/* Wait for all interfaces to be ready */
+		if (n_assigned != n_reserved) {
+			link_data->reserved_ready = true;
+			err = 0;
+			goto out;
+		}
+
+		for_each_sdata_link(local, link) {
+			struct ieee80211_link_data *tmp;
+
+			for_each_valid_link(&link->sdata->wdev, link_id) {
+				tmp = sdata_dereference(link->sdata->link[link_id],
+							link->sdata);
+				if (!tmp)
+					continue;
+				wiphy_work_queue(link->sdata->local->hw.wiphy,
+						 &tmp->csa.finalize_work);
+			}
+		}
 	}
 
 out:
-- 
2.45.2

